src/radix_tree.cpp

```diff
@@ -0,0 +1,602 @@
+
+node::node(unsigned char* data)
+    : data_(data)
+{}
+
+inline
+uint32_t node::refcount()
+{
+    uint32_t u32;
+    memcpy(&u32, data_, sizeof(u32));
+    return u32;
+}
+
+inline
+void node::set_refcount(uint32_t value)
+{
+    memcpy(data_, &value, sizeof(value));
+}
+
+inline
+uint32_t node::prefix_length()
+{
+    uint32_t u32;
+    memcpy(&u32, data_ + sizeof(uint32_t), sizeof(u32));
+    return u32;
+}
+
+inline
+void node::set_prefix_length(uint32_t value)
+{
+    memcpy(data_ + sizeof(value), &value, sizeof(value));
+}
+
+inline
+uint32_t node::edgecount()
+{
+    uint32_t u32;
+    memcpy(&u32, data_ + 2 * sizeof(uint32_t), sizeof(u32));
+    return u32;
+}
+
+inline
+void node::set_edgecount(uint32_t value)
+{
+    memcpy(data_ + 2 * sizeof(value), &value, sizeof(value));
+}
+
+inline
+unsigned char* node::prefix()
+{
+    return data_ + 3 * sizeof(uint32_t);
+}
+
+inline
+void node::set_prefix(unsigned char const* bytes)
+{
+    memcpy(prefix(), bytes, prefix_length());
+}
+
+inline
+unsigned char* node::first_bytes()
+{
+    return prefix() + prefix_length();
+}
+
+inline
+void node::set_first_bytes(unsigned char const* bytes)
+{
+    memcpy(first_bytes(), bytes, edgecount());
+}
+
+inline
+unsigned char node::first_byte_at(size_t i)
+{
+    zmq_assert(i < edgecount());
+    return first_bytes()[i];
+}
+
+inline
+void node::set_first_byte_at(size_t i, unsigned char byte)
+{
+    zmq_assert(i < edgecount());
+    first_bytes()[i] = byte;
+}
+
+inline
+unsigned char* node::node_ptrs()
+{
+    return prefix() + prefix_length() + edgecount();
+}
+
+inline
+void node::set_node_ptrs(unsigned char const* ptrs)
+{
+    memcpy(node_ptrs(), ptrs, edgecount() * sizeof(void*));
+}
+
+inline
+node node::node_at(size_t i)
+{
+    zmq_assert(i < edgecount());
+
+    unsigned char* data;
+    memcpy(&data, node_ptrs() + i * sizeof(void*), sizeof(data));
+    return node(data);
+}
+
+inline
+void node::set_node_at(size_t i, node n)
+{
+    zmq_assert(i < edgecount());
+    memcpy(node_ptrs() + i * sizeof(void*),
+                &n.data_,
+                sizeof(n.data_));
+}
+
+inline
+void node::set_edge_at(size_t i, unsigned char byte, node n)
+{
+    set_first_byte_at(i, byte);
+    set_node_at(i, n);
+}
+
+inline
+bool node::operator==(node other) const
+{
+    return data_ == other.data_;
+}
+
+inline
+bool node::operator!=(node other) const
+{
+    return !(*this == other);
+}
+
+void node::resize(size_t prefix_length, size_t edgecount)
+{
+    size_t sz = 3 * sizeof(uint32_t) + prefix_length
+        + edgecount * (1 + sizeof(void*));
+    unsigned char* new_data = static_cast<unsigned char*>(realloc(data_, sz));
+    zmq_assert(new_data);
+    data_ = new_data;
+    set_prefix_length(static_cast<uint32_t>(prefix_length));
+    set_edgecount(static_cast<uint32_t>(edgecount));
+}
+
+node make_node(size_t refs, size_t bytes, size_t edges)
+{
+    size_t size = 3 * sizeof(uint32_t) + bytes
+        + edges * (1 + sizeof(void*));
```

Comment:
Maybe define a struct for the "header" part? This would significantly simplify size calculations and field accesses.

Reply:
Correct me if I got this wrong, would it look something like this?

```
struct header { /* 3 uint32_ts here */ };

struct node
{
    header hdr;
    unsigned char *data;
};
```

I'm not sure how this would affect the data layout. AFAIK the cleaner way would be to use C99's flexible array members, but that isn't standard C++. Using flexible array members will change `node` to:

```
struct node
{
    uint32_t refcount;
    uint32_t prefix_length;
    uint32_t edgecount;
    unsigned char data[];
};
```

This would eliminate the getter/setter functions for the first 3 fields with no change in the data layout.

cc @bjovke

Reply:
@ssbl Flexible array members are not standard in C++.
One of the ways to do this is to preserve the `data_` member as it is and cast it to `struct node` when needed.
You might consider making `data_` void pointer since it's already being casted many times. It would be consistent with `malloc()`/`free()`.
This would make the code more verbose but also clearer and less error prone.
Also, whenever pointer arithmetic is being done it's a good idea to put some comments.

For example:

```
struct node
{
    // Header is located at the beginning of data_ block.
    struct header
    {
        uint32_t refcount;
        uint32_t prefix_length;
        uint32_t edgecount;
    }

    void *data_;
};
```

`make_node()` might look like this:

```
node make_node (size_t refs, size_t bytes, size_t edges)
{
    node n (malloc (sizeof (node::header) + bytes + edges * (1 + sizeof (void *))));
    zmq_assert (n.data_);

    node::header &hdr = *reinterpret_cast<node::header *>(n.data_);
    hdr.refcount = refs;
    hdr.prefix_length = bytes;
    hdr.edgecount = edges;

    return n;
}
```
prefix_length() might look like this:

```
uint32_t node::prefix_length ()
{
    return reinterpret_cast<node::header *>(data_)->prefix_length;
}
```

Reply:
@bjovke Thanks, that makes sense!